src/to-be-labelled.js

```diff
@@ -0,0 +1,124 @@
+import {matcherHint, printReceived} from 'jest-matcher-utils'
+import {checkHtmlElement, getTag} from './utils'
+
+/*
+ * Lists all tags accepting alt as an attribute
+ * @link https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
+ */
+const TAGS_WITH_ALT = ['applet', 'area', 'img']
+
+const TAGS_WITH_TITLE = ['svg']
+
+const TAGS_WITH_TITLE_ATTR = ['input', 'select', 'textarea', 'abbr']
+
+const TAGS_IMAGES = ['img', 'svg']
+
+const TAGS_WITH_TEXT_CONTENT = ['a', 'button', 'object']
+
+/*
+ * Checks if the element is labelled by `aria-label`
+ * @link https://www.w3.org/TR/WCAG20-TECHS/ARIA6.html
+ */
+function isHavingARIALabel(element) {
+  return element.hasAttribute('aria-label')
+}
+
+/*
+ * Checks if the element is labelled by `aria-labelledby`
+ * @link https://www.w3.org/TR/WCAG20-TECHS/ARIA10.html
+ */
+function isHavingARIALabelledBy(element) {
+  return element.hasAttribute('aria-labelledby')
+}
+
+/*
+ * Checks if meaningful images are having an alt attribute
+ * @link https://www.w3.org/TR/WCAG20-TECHS/H37.html
+ */
+function isHavingMeaningfulAlt(element) {
+  return (
+    TAGS_WITH_ALT.includes(getTag(element)) &&
+    element.hasAttribute('alt') &&
+    element.getAttribute('alt') !== ''
+  )
+}
+
+/*
+ * Checks if there is a title tag as a child
+ * @link https://www.w3.org/TR/WCAG20-TECHS/H65.html
+ */
+function isHavingMeaningfulTitle(element) {
+  return (
+    TAGS_WITH_TITLE.includes(getTag(element)) &&
+    element.querySelector('title') !== null
+  )
+}
+
+function isHavingMeaningfulTitleAttribute(element) {
+  return (
+    TAGS_WITH_TITLE_ATTR.includes(getTag(element)) &&
+    element.hasAttribute('title') &&
+    element.getAttribute('title') !== null
+  )
+}
+
+/*
+ * Checks if there is an adjacent image and text
+ * @link https://www.w3.org/TR/WCAG20-TECHS/G95.html
+ */
+function isHavingASiblingLabel(element) {
+  return (
+    TAGS_IMAGES.includes(getTag(element)) &&
+    TAGS_WITH_TEXT_CONTENT.includes(getTag(element.parentElement)) &&
+    element.parentElement.textContent !== ''
+  )
+}
+
+/*
+ * Checks if there is an adjacent image and text
+ * @link https://www.w3.org/TR/WCAG20-TECHS/H2.html
+ */
+function isHavingAContentText(element) {
+  const isHavingLabelledImg =
+    element.querySelector('img') &&
+    (isHavingARIALabel(element.querySelector('img')) ||
+      isHavingARIALabelledBy(element.querySelector('img')) ||
+      isHavingMeaningfulAlt(element.querySelector('img')))
+
+  const isHavingLabelledSVG =
+    element.querySelector('svg') &&
+    (isHavingARIALabel(element.querySelector('svg')) ||
+      isHavingARIALabelledBy(element.querySelector('svg')) ||
+      isHavingMeaningfulTitle(element.querySelector('svg')))
+
+  return (
+    TAGS_WITH_TEXT_CONTENT.includes(getTag(element)) &&
+    (element.textContent !== '' || isHavingLabelledImg || isHavingLabelledSVG)
+  )
+}
+
+export function toBeLabelled(element) {
```

Comment:
Nice idea! For me, I had more in mind the idea of checking if there is an accessible name for a targetted element, but this could be interesting.

In this case, imo, we should implement also the same kind of calculation as the one described by the [Accessibility API implementation guide](https://www.w3.org/TR/html-aapi/#accessible-name-and-description-calculation).

This is mainly in case there is several attributes on one element like: 
```html
<img src="" alt="Text content" aria-label="Accessible name" />
```

In this specific case, the final text content read by assistive technologies would be _Accessible name_ and not _Text content_. But in case the image is not available, the browser may show the `alt` content, so _text content_.
What is your take on this, what should `.toBeLabelled` check and consider as `true`?

Reply:
Not sure if I am still comfortable doing this as part of this PR. Could we create an issue for this specific point and having it part of another PR. I may need help for this specific part :)

Reply:
Given that it will not involve a breaking change in the future, I think that's perfectly reasonable.